### PR TITLE
ui: add badges to records

### PR DESCRIPTION
**Pull request by TLGINO**
* closes https://github.com/inveniosoftware/invenio-app-rdm/issues/2020

### CODE REVIEW

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (ntarocco) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:15]:
```suggestion

```

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (ntarocco) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:30]: Do we want to have mandatory `svg`? Or can it be a `jpg` or `png`? Maybe better to add it to the `resource.icon` config.

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (ntarocco) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:37]: When using `target="_blank"`, I would also add [noreferrer](https://stackoverflow.com/questions/50773152/when-should-i-use-rel-noreferrer) (check if there are other occurrences).

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (ntarocco) [inline invenio_app_rdm/records_ui/views/deposits.py:118]:
```suggestion
            "APP_RDM_RECORD_LANDING_PAGE_EXTERNAL_LINKS", []
```

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (ntarocco) [inline invenio_app_rdm/records_ui/views/deposits.py:93]: Does the `scheme` allows providing a subclass of `ResourceParser` to override the default? I am not sure that I understand how it can be used :)/

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (TLGINO) [reply invenio_app_rdm/records_ui/views/deposits.py:115]: It gets called here:

https://github.com/TLGINO/invenio-app-rdm/blob/1aae1a016e2e79e3c6681a0379df28aaf1dfd057/invenio_app_rdm/records_ui/views/deposits.py#L96

```
if hasattr(ResourceParser, scheme):
    parser = ResourceParser(resource)
    item = getattr(parser, scheme)()
```

This is done with the idea that we may want to parse different types of `schema` and not just `url`
In which caase it would suffice to add another method to `ResourceParser` with a name corresponding to it's `schema`

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (TLGINO) [reply invenio_app_rdm/records_ui/views/deposits.py:93]: This is also in the above answer but I'll link an image here for clarity

image

The idea is that we would be able to parse not only `url` but the others too if we want (thus avoiding a bunch of ugly if chaining)
In retrospect this may be a bit overkill :sweat_smile:

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (TLGINO) [reply invenio_app_rdm/records_ui/views/deposits.py:142]: I'm waiting for a review by @slint first as a demo of how the parser would work
But yes these aren't meant to be merged (and could be put in the docs)

File: invenio_app_rdm/config.py
Comment (TLGINO) [reply invenio_app_rdm/config.py:685]: sure!

**Review by slint — COMMENTED** [review]
The pieces for the feature are there, but there are some major changes needed to make things go in the right direction of what's the final API/config we want to expose.

File: invenio_app_rdm/records_ui/views/deposits.py
Comment (slint) [reply invenio_app_rdm/records_ui/views/deposits.py]: I would delegate all the complexity/freedom of generating the entry to these functions (or classes). e.g. if someone wants to render an entry for an EAN13 related identifier, it's completely up to them how they do it, and we don't do any validation or checks.

**Comment by TLGINO** [comment]
PS: PR is up for style / design review, not so much for actual implementation (ie ignore the images + dummy title / description names)

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (kpsherva) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:31]: I would check if the path is accessible by some variables in jinja (there was a way to access statics if I remember correctly), instead of hardcoding it

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (kpsherva) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:38]: please don't set inline styling, these should live under the theme files. In some cases inline styling will be triggering content security policy errors.
There are already some classes defined to cover this styling. Check https://github.com/inveniosoftware/invenio-theme/tree/master/invenio_theme/assets/semantic-ui/less/invenio_theme/theme (global theme) 
and https://github.com/inveniosoftware/invenio-app-rdm/tree/master/invenio_app_rdm/theme/assets/semantic-ui/less/invenio_app_rdm/theme

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (kpsherva) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:13]: see the comment below about inline styling

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (kpsherva) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:25]: I would use semantic ui list element https://react.semantic-ui.com/elements/list/#content-image (in pure html version, not react), which is already styled to what you need

**Review by kpsherva — COMMENTED** [review]
I left a couple of comments about styling mainly :)

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (TLGINO) [reply invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:31]: Do you mean the 
`url_for`
method?
Like this:

` {{ license.id }} icon`

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (TLGINO) [reply invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:38]: Ok I'll fix it!
Thx for the review also! :)

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (kpsherva) [reply invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:28]: please check: https://github.com/inveniosoftware/invenio-theme/blob/master/invenio_theme/templates/semantic-ui/invenio_theme/header.html#L25

File: invenio_app_rdm/records_ui/views/records.py
Comment (kpsherva) [inline invenio_app_rdm/records_ui/views/records.py:206]: you need to add it to the context of request views as well, they use the record template

**Review by kpsherva — APPROVED** [review]

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [inline invenio_app_rdm/records_ui/utils.py:83]:
```suggestion
def external_resource_dumper(identifier, icon, title, subtitle=None, section=None, template=None):
    """Dumps the data to be rendered into a dictionary."""
    return {
        "content": {
            "url": identifier.get("identifier"),
            "title": title,
            "subtitle": subtitle,
            "icon": url_for("static", filename="images/" + icon),
            "section": section,
        },
        "template": template,
    }
```

minor: to be able to pass a `template` if needed

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (slint) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:13]: minor: I think we have to make `section` a required parameter/field, just to avoid all the issues with how/were to display the ones that have `section: None`.

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [inline invenio_app_rdm/records_ui/utils.py:62]:
```suggestion
def get_external_resources(record):
    """Generate external resource links for a record."""
```

nit: naming and docstring

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [inline invenio_app_rdm/records_ui/utils.py:69]: nit: I somewhat prefer [the `itertools.chain.from_iterable` approach ](https://stackoverflow.com/a/953097), instead of the nested list comprehensions in terms of readability.

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [inline invenio_app_rdm/records_ui/utils.py:73]:
```suggestion
def dump_external_resource(identifier, icon, title, subtitle=None, section=None):
    """Dumps a external resource dictionary."""
```

nit: naming to follow the convention of "verbs for functions"/"things for classes" and docstring

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [inline invenio_app_rdm/records_ui/utils.py:79]:
```suggestion
            "icon": icon,
```

mionr: I would delegate the `url_for` call to the calling functions (e.g. in case someone decides to pass a link to CDN)

File: invenio_app_rdm/requests_ui/views/requests.py
Comment (slint) [inline invenio_app_rdm/requests_ui/views/requests.py:128]: minor: oof, is this needed for this specific template? I know there's another template for previewing a record, but I'm not sure where exactly it is... I would ask someone familiar with deposit form/UI to get the right place.

**Review by slint — COMMENTED** [review]
Nice one, looks pretty good! Some naming/docs nits and minor things, but otherwise we're good to fly.

Can you also update/share a screenshot with some examples? To generate them quickly without having to implement the function logic, I would just hardcode the values.

File: invenio_app_rdm/requests_ui/views/requests.py
Comment (kpsherva) [reply invenio_app_rdm/requests_ui/views/requests.py:128]: it is needed, I asked to include it there, otherwise the request's record tab is not displayed properly with full info

**Comment by kpsherva** [comment]
also a rebase needed :)

File: invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html
Comment (ppanero) [inline invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/external_resources.html:2]:
```suggestion
Copyright (C) 2023 CERN.
```

File: invenio_app_rdm/config.py
Comment (ppanero) [inline invenio_app_rdm/config.py:710]: I'm not a fan of such long names, but I do not have a better one in mind :/

**Review by ppanero — COMMENTED** [review]
Very good work! Just very tiny comments or doubts (not blocking merge)

File: invenio_app_rdm/records_ui/utils.py
Comment (TLGINO) [reply invenio_app_rdm/records_ui/utils.py:74]: Because we could potentially have 3 urls to "render"
2 of them could be github
And 1 openaire
The weirdness (no workaround) is that the renderer say for github will take the 3 urls
And then will output 2 formated dicts
The renderer for openaire will also take the 3 urls
And will only output 1 formated dict
Hence in the end we get something like [[gh1, gh2], [openaire1]] which we need to unnest

File: invenio_app_rdm/records_ui/utils.py
Comment (slint) [reply invenio_app_rdm/records_ui/utils.py:72]: The issue is that the passed `render` functions just receive a record, but are free to do any kind of manipulation they want (could e.g. make service calls to fetch additional info). If this was a clear-cut API in terms of accessed layers/APIs, we could narrow it down, but for now, it acts as a fail-safe so that we don't 500 on rendering the record landing page.

**Pull request merged**
